### PR TITLE
fix: bom stuff

### PR DESCRIPTION
**Pull request by paperclover**
pro tip: do not merge PRs that do not have tests

note to self: do not open PRs marked as "ready to review" without tests and also forget to mention in the notes that there are not tests

### CODE REVIEW

File: src/bun.js/webcore/blob.zig [hunk 1]
```diff
@@ -4488,27 +4482,45 @@ pub const Blob = struct {
             // we don't need to clone
             .clone => {
                 this.store.?.ref();
+
+                if (bom == .utf16_le) {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/blob.zig:4549]: memory leak right here

File: src/bun.js/webcore/blob.zig [hunk 2]
```diff
@@ -4488,27 +4482,45 @@ pub const Blob = struct {
...
+                }
                 // we don't need to worry about UTF-8 BOM in this case because the store owns the memory.
                 return ZigString.init(buf).external(global, this.store.?, Store.external);
             },
             .transfer => {
                 const store = this.store.?;
                 std.debug.assert(store.data == .bytes);
                 this.transfer();
+
+                if (bom == .utf16_le) {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/blob.zig:4562]: memory leak right here

File: src/bun.js/webcore/blob.zig [hunk 3]
```diff
@@ -4488,27 +4482,45 @@ pub const Blob = struct {
...
                 // we don't need to worry about UTF-8 BOM in this case because the store owns the memory.
                 return ZigString.init(buf).external(global, store, Store.external);
             },
             // strings are immutable
             // sharing isn't really a thing
             .share => {
                 this.store.?.ref();
-                // we don't need to worry about UTF-8 BOM in this case because the store owns the memory.s
+
+                if (bom == .utf16_le) {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/blob.zig:4575]: memory leak right here

File: src/bun.js/webcore/blob.zig [hunk 4]
```diff
@@ -4518,6 +4530,12 @@ pub const Blob = struct {
                     return out.toJS(global);
                 }
 
+                if (bom == .utf16_le) {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/blob.zig:4597]: and memory leak right here

File: src/bun.js/webcore/blob.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/webcore/blob.zig:4549]: move this above this.store.?.ref() to fix the leak

File: src/bun.js/webcore/blob.zig (hunk 2 above)
Comment (Jarred-Sumner) [reply src/bun.js/webcore/blob.zig:4562]: call this.detach() before this.transfer() to fix the leak

File: src/bun.js/webcore/blob.zig [hunk 5]
```diff
@@ -4518,6 +4531,12 @@ pub const Blob = struct {
...
+                    var out = bun.String.createUTF16(bun.reinterpretSlice(u16, buf));
+                    defer out.deref();
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/blob.zig:4599]: this still leaks!

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
1 more leak

File: src/bun.js/webcore/blob.zig (hunk 1 above)
Comment (jdalton) [reply src/bun.js/webcore/blob.zig:4549]: Learning moment for me. How is this a leak and why does moving it above fix it?
